base/compiler/utilities.jl

```diff
@@ -133,6 +133,8 @@ is_nospecializeinfer(method::Method) = method.nospecializeinfer && is_nospeciali
 # MethodInstance/CodeInfo #
 ###########################
 
+MethodInstance() = ccall(:jl_new_method_instance_uninit,Ref{MethodInstance},())
```

Comment:
This is probably unsound, since we declare (in jltypes.c) that this constructor is not permitted to exist

Reply:
Thanks for catching this and trying to explain why it's a problem. I don't follow why it's unsound.

What do you mean by "unsound"? Is it that calling this function produces an incomplete method instance? Or is it like misuse of `@assume_effects :nothrow` where declaring the method itself is UB?

How is that this constructor is not permitted to exist declared? I searched src/jltypes.c for "unsound", "MethodInstance", and "jl_new_method_instance_uninit" and found nothing.

If this is unsound, then are the existing usages of `ccall(:jl_new_method_instance_uninit,Ref{MethodInstance},())` also unsound?

Reply:
When you create a new datatype, you specify how many of its fields are guaranteed to be initialized. 

I assume Jameson refers to this as breaking the invariant. The existing uses are indeed "unsound", but they are exclusively in the tests and not part of the compiler interface.